File: meudin_app/lib/db/transaction_dao.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:meudin_app/models/forms/new_revenue_form.dart';
import 'package:meudin_app/models/smart_error.dart';
import 'package:meudin_app/services/smart_error_service.dart';
import 'package:meudin_app/utils/utils.dart';

class TransactionDao {
  final dbReference = FirebaseFirestore.instance;
  final String _transactionCollectionName = 'transactions';

  late SmartErrorService _smartErrorService;

  TransactionDao() {
    _smartErrorService = SmartErrorService();
  }
  Future<String> saveNewRevenue(NewRevenueForm form) async {
    Timestamp payDay = Timestamp.fromDate(form.payDay);
    Timestamp creationDate = Timestamp.fromDate(DateTime.now());

    String? uid = form.uid;
    String? reason = form.reason.text;
    double amount = Utils.convertStringFormToDouble(form.revenueValue.text);
    String? walletId = form.walletId;

    var batch = dbReference.batch();

    var docRef = dbReference.collection(_transactionCollectionName).doc();

    String id = docRef.id;

    batch.set(docRef, {
      'id': id,
      'userId': uid,
      'reason': reason,
      'transactionDate': payDay,
      'walletId': walletId,
      'amount': amount,
      'type': 'REVENUE',
      'creationDate': creationDate,
    });

    await batch.commit().then((value) {
      return id;
    }).catchError((onError) {
      SmartError errorDto = SmartError();
      errorDto.errorData = onError;
      errorDto.errorLog = onError.toString();
      errorDto.feature = 'Save revenue';
      errorDto.userId = id;
      errorDto.userMail = '';

      _smartErrorService.saveError(errorDto);
      return '';
    });

    return id;
  }
}

File: API/functions/src/db/transaction-dao.ts
import { Transaction } from "../models/transaction";
import { db } from "../index";
import { NewWasteDto } from "../models/dtos/new-waste-dto";
import { NewRevenueDto } from "../models/dtos/new-revenue-dto";

export class TransactionDao {
    transactionCollectionName = Transaction.collectionName

    async saveNewWaste(newWasteDto: NewWasteDto) {

        const batch = db.batch()

        const transactionDocRef = db.collection(this.transactionCollectionName).doc()

        var id = transactionDocRef.id

        let now = new Date()

        batch.set(transactionDocRef, {
            'id': id,
            'userId': newWasteDto.uid,
            'reason': newWasteDto.reason,
            'transactionDate': newWasteDto.spendDate,
            'walletId': newWasteDto.walletId,
            'amount': newWasteDto.waste,
            'categoryId': newWasteDto.categoryId,
            'type': 'WASTE',
            'creationDate': now,
        })

        await batch.commit()

        return id
    }

    async saveNewRevenue(newRevenueDto: NewRevenueDto) {

        const batch = db.batch()

        const transactionDocRef = db.collection(this.transactionCollectionName).doc()

        var id = transactionDocRef.id

        let now = new Date()

        batch.set(transactionDocRef, {
            'id': id,
            'userId': newRevenueDto.uid,
            'reason': newRevenueDto.reason,
            'transactionDate': newRevenueDto.payDay,
            'walletId': newRevenueDto.walletId,
            'amount': newRevenueDto.amount,
            'type': 'REVENUE',
            'creationDate': now,
        })

        await batch.commit()

        return id
    }

    async getTransactionsByWalletIdAndDateIntervalRes(walletId: string, startDate: Date, endDate: Date): Promise<Transaction[]> {
        var transactionList: Transaction[] = []

        var snapshot = await db
            .collection(this.transactionCollectionName)
            .where('walletId', '==', walletId)
            .where('transactionDate', '>=', startDate)
            .where('transactionDate', '<=', endDate)
            .get()

        if (snapshot.empty) {
            return transactionList
        } else {
            var transaction

            for (const doc of snapshot.docs) {
                transaction = new Transaction(doc)

                transactionList.push(transaction)
            }
        }

        return transactionList
    }
}